scarlet/source.py

```diff
@@ -3,344 +3,182 @@
-
-    Returns
-    -------
-    None.
-    But `source.sed` and `source.morph` are set.
-    """
-    assert source.K==1
-    B = img.shape[0]
-    source.sed[0] = get_peak_sed(img, source.center_int)
-    _y, _x = source.center_int
-    Ny = 2*min(img.shape[1]-_y, _y)-1
-    Nx = 2*min(img.shape[2]-_x, _x)-1
-
-    # If the source is on the edge, extend the image with its reflection
-    # to model the hidden portion in x and/or y
-    if Ny<=1:
-        _Ny = img.shape[1]
-        Ny = 2*img.shape[1]-1
-        _img = np.zeros((img.shape[0], Ny, img.shape[2]))
-        _img[:,:_Ny-1] = np.fliplr(img[:,1:])
-        _img[:,_Ny-1:] = img[:]
-        _y = _Ny - 1
-    else:
-        _img = img.copy()
-    if Nx<=1:
-        _Nx = img.shape[2]
-        Nx = 2*img.shape[2]-1
-        __img = _img.copy()
-        _img = np.zeros((_img.shape[0], _img.shape[1], Nx))
-        _img[:,:,:_Nx-1] = np.fliplr(__img[:,:,1:])
-        _img[:,:,_Nx-1:] = __img
-        _x = _Nx - 1
-    cx, cy = Nx // 2, Ny // 2
-
-    morph = np.zeros((Ny,Nx))
-    # use the band with maximum flux for the source
-    band = np.argmax(_img[:,_y,_x])
-    morph[:] = _img[band,_y-cy:_y+cy+1,_x-cx:_x+cx+1]/source.sed[0,band]
-    morph = morph.reshape((morph.size,))
-    morph[morph<0] = 0
-    # Apply the appropriate constraints
+    center_pix = morph.size // 2
+    morph[center_pix] = max(img[:,_y,_x].sum(axis=0), tiny)
+    return sed.reshape((1,B)), morph.reshape((1, shape[0], shape[1]))
+
+def init_above_noise(source, img, bg_rms, thresh=1., symmetric=True, monotonic=True):
+
+    # every source as large as the entire image, but shifted to its centroid
+    B, Ny, Nx = img.shape
+    source._set_frame(source.center, (Ny,Nx))
+
+    # determine initial SED from peak position
+    try:
+        sed = get_pixel_sed(img, source.center_int)
+    except SourceInitError:
+        # flat weights as fall-back
+        sed = np.ones(B) / B
+
+    # build optimal detection coadd
+    weights = np.array([sed[b]/bg_rms[b]**2 for b in range(B)])
+    jacobian = np.array([sed[b]**2/bg_rms[b]**2 for b in range(B)]).sum()
+    detect = np.einsum('i,i...', weights, img) / jacobian
+
+    # copy morph from detect cutout, make non-negative
+    source_slice = source.get_slice_for(img.shape)
+    morph = np.zeros((source.Ny, source.Nx))
+    morph[source_slice[1:]] = detect[source.bb[1:]]
+
+    # check if source_slice is covering the whole of morph:
+    # if not, extend morph by coping last row/column from img
+    if source_slice[1].stop - source_slice[1].start < source.Ny:
+        morph[0:source_slice[1].start,:] = morph[source_slice[1].start,:]
+        morph[source_slice[1].stop:,:] = morph[source_slice[1].stop-1,:]
+    if source_slice[2].stop - source_slice[2].start < source.Nx:
+        morph[:,0:source_slice[2].start] = morph[:,source_slice[2].start][:,None]
+        morph[:,source_slice[2].stop:] = morph[:,source_slice[2].stop-1][:,None]
```

Comment:
Yes, but the source (in its own box) would be truncated at the edge of the image if it isn't extended. If not, and you then hit it with symmetry, the inside part of the source will be truncated as well (because the outside is zero). In this case, the outside part is identical to the value of the edge column/row.

Granted, I'm assuming that symmetry is used, which is the default. In principle, we might make it hard-coded, so that this initialization behaves properly.

Reply:
Yeah, I also wasn't paying attention. Since `source.Ny`,` source.Nx` are the same as the image shape, I think that this is fine. For some reason I was (incorrectly) thinking that `source.Ny`,`source.Nx` could be different, and that's what I was objecting to.